### PR TITLE
Submission Service - Enhance TEK's 

### PR DESCRIPTION
**Pull request by ioangut**
## Description

* This PR is the implementation for [EXPOSUREBACK-54.](https://jira.itc.sap.com/browse/EXPOSUREBACK-54).
* Enhance submission service to accept duplicated TEK's and flexible rolling period.

### CODE REVIEW

File: common/persistence/src/test/java/app/coronawarn/server/common/persistence/domain/DiagnosisKeyBuilderTest.java [hunk 1]
```diff
@@ -165,17 +165,17 @@ void transmissionRiskLevelDoesNotThrowForValid(int validRiskLevel) {
...
         .isInstanceOf(InvalidDiagnosisKeyException.class)
-        .hasMessage("[Rolling period must be " + DiagnosisKey.EXPECTED_ROLLING_PERIOD
+        .hasMessage("[Rolling period must be between "+ DiagnosisKey.MIN_ROLLING_PERIOD + " and " + DiagnosisKey.MAX_ROLLING_PERIOD
             + ". Invalid Value: " + invalidRollingPeriod + "]");
   }
 
   @Test
   void rollingPeriodDoesNotThrowForValid() {
-    assertThatCode(() -> keyWithRollingPeriod(DiagnosisKey.EXPECTED_ROLLING_PERIOD)).doesNotThrowAnyException();
+    assertThatCode(() -> keyWithRollingPeriod(DiagnosisKey.MAX_ROLLING_PERIOD)).doesNotThrowAnyException();
```
Comment (MichaelPFrey) [inline common/persistence/src/test/java/app/coronawarn/server/common/persistence/domain/DiagnosisKeyBuilderTest.java:178]: should there be more test? Like for example for `DiagnosisKey.MIN_ROLLING_PERIOD` and a value between `MIN` and `MAX`?

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java [hunk 2]
```diff
@@ -86,10 +87,17 @@ public SubmissionPayloadValidator(SubmissionServiceConfig submissionServiceConfi
     public boolean isValid(SubmissionPayload submissionPayload, ConstraintValidatorContext validatorContext) {
       List<TemporaryExposureKey> exposureKeys = submissionPayload.getKeysList();
       validatorContext.disableDefaultConstraintViolation();
-
-      boolean isValid = checkKeyCollectionSize(exposureKeys, validatorContext);
-      isValid &= checkUniqueStartIntervalNumbers(exposureKeys, validatorContext);
-      isValid &= checkNoOverlapsInTimeWindow(exposureKeys, validatorContext);
+      boolean isValid = false;
+
+      for (TemporaryExposureKey exposureKey: exposureKeys) {
```
Comment (EugenM-SAP) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:92]: Something doesn't look right here. For each exposure key in the payload we check the constraints of the entire set? I'm referring to the else branch. Can you confirm that is intended?

File: services/submission/src/test/resources/application.yaml [hunk 3]
```diff
@@ -33,6 +33,8 @@ services:
       path: /version/v1/tan/verify
     monitoring:
       batch-size: 5
+    diagnosis-key:
+      rolling-period: 144
```
Comment (EugenM-SAP) [inline services/submission/src/test/resources/application.yaml:37]: Does this configuration represent the maximum rolling period? If yes, it should be evident from the name

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java [hunk 4]
```diff
@@ -132,16 +136,45 @@ private boolean checkNoOverlapsInTimeWindow(List<TemporaryExposureKey> exposureK
 
       Integer[] sortedStartIntervalNumbers = exposureKeys.stream()
           .mapToInt(TemporaryExposureKey::getRollingStartIntervalNumber)
-          .sorted().boxed().toArray(Integer[]::new);
+          .distinct().sorted().boxed().toArray(Integer[]::new);
```
Comment (pithumke) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:139]: This breaks the validation for TEK sets without flexible rolling periods, as it removed duplicate rolling start interval numbers. With some refactoring, `checkUniqueStartIntervalNumbers` could be removed altogether (see above).

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java [hunk 5]
```diff
@@ -86,11 +87,14 @@ public SubmissionPayloadValidator(SubmissionServiceConfig submissionServiceConfi
     public boolean isValid(SubmissionPayload submissionPayload, ConstraintValidatorContext validatorContext) {
       List<TemporaryExposureKey> exposureKeys = submissionPayload.getKeysList();
       validatorContext.disableDefaultConstraintViolation();
```
Comment (pithumke) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:89]: Consider adding a test which ensures that all rolling start interval numbers are at 00:00 UTC. This way, the `checkNoOverlapsInTimeWindow` test can be skipped for TEK sets without flexible rolling periods.

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java [hunk 6]
```diff
@@ -86,11 +87,14 @@ public SubmissionPayloadValidator(SubmissionServiceConfig submissionServiceConfi
...
 
-      boolean isValid = checkKeyCollectionSize(exposureKeys, validatorContext);
-      isValid &= checkUniqueStartIntervalNumbers(exposureKeys, validatorContext);
-      isValid &= checkNoOverlapsInTimeWindow(exposureKeys, validatorContext);
-
+      if (keysHaveFlexibleRollingPeriod(exposureKeys)) {
+        isValid &= checkKeysCummulateEqualOrLessThanMaxRollingPeriodPerDay(exposureKeys, validatorContext);
+      } else {
+        isValid &= checkKeyCollectionSize(exposureKeys, validatorContext);
+        isValid &= checkUniqueStartIntervalNumbers(exposureKeys, validatorContext);
```
Comment (pithumke) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:96]: Can be put into one step with `checkNoOverlapsInTimeWindow`

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java [hunk 7]
```diff
@@ -132,16 +136,45 @@ private boolean checkNoOverlapsInTimeWindow(List<TemporaryExposureKey> exposureK
...
+
+    private boolean checkKeysCummulateEqualOrLessThanMaxRollingPeriodPerDay(List<TemporaryExposureKey> exposureKeys,
+        ConstraintValidatorContext validatorContext) {
+
+      HashMap<Integer, Integer> totalKeysPerDay = new HashMap<>();
+
+      for (TemporaryExposureKey exposureKey : exposureKeys) {
+        int numberOfKeys = exposureKey.getRollingPeriod();
+        if (totalKeysPerDay.containsKey(exposureKey.getRollingStartIntervalNumber())) {
+          numberOfKeys += totalKeysPerDay.get(exposureKey.getRollingStartIntervalNumber());
```
Comment (pithumke) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:159]: I'm not sure if this is testing the correct constraint. AFAIK, the rolling period sums of all TEKs with the same rolling start interval number must be <= 144. But here, the number of keys are being counted?

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java [hunk 8]
```diff
@@ -132,16 +136,45 @@ private boolean checkNoOverlapsInTimeWindow(List<TemporaryExposureKey> exposureK
...
       }
       return true;
     }
+
+    private boolean checkKeysCummulateEqualOrLessThanMaxRollingPeriodPerDay(List<TemporaryExposureKey> exposureKeys,
+        ConstraintValidatorContext validatorContext) {
+
+      HashMap<Integer, Integer> totalKeysPerDay = new HashMap<>();
+
+      for (TemporaryExposureKey exposureKey : exposureKeys) {
```
Comment (pithumke) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:156]: Consider grouping by start interval number, summing up the rolling periods for each group and then checking if `anyMatch` for sum > 144. Should be a lot shorter and more memory efficient.

File: services/submission/src/test/java/app/coronawarn/server/services/submission/controller/PayloadValidationTest.java [hunk 9]
```diff
@@ -109,30 +110,113 @@ void check200ResponseStatusForGapsInTimeIntervals() {
...
+  void check200ResponseStatusForMoreThan14KeysWithValidFlexibleRollingPeriod() {
+    ResponseEntity<Void> actResponse = executor.executePost(buildPayloadWithMoreThan14KeysAndFlexibleRollingPeriod());
+    assertThat(actResponse.getStatusCode()).isEqualTo(OK);
+  }
+
+  private Collection<TemporaryExposureKey> buildPayloadWithMoreThan14KeysAndFlexibleRollingPeriod() {
+    ArrayList<TemporaryExposureKey> flexibleRollingPeriodKeys = new ArrayList<>();
+    int counter = 0;
+    /* Generate keys with fixed rolling period (144) for the past 20 days */
+    for ( ; counter <= 20; counter++) {
```
Comment (pithumke) [inline services/submission/src/test/java/app/coronawarn/server/services/submission/controller/PayloadValidationTest.java:156]: Introduce separate counters for each loop for improved readability.

File: services/submission/src/test/java/app/coronawarn/server/services/submission/controller/PayloadValidationTest.java [hunk 10]
```diff
@@ -109,30 +110,113 @@ void check200ResponseStatusForGapsInTimeIntervals() {
...
+          createRollingStartIntervalNumber(2) - counter * DiagnosisKey.MAX_ROLLING_PERIOD, 3, 133));
+
+    }
+    return flexibleRollingPeriodKeys;
+  }
 
-    ResponseEntity<Void> actResponse = executor.executePost(keysWithValidStartIntervalNumber);
+  @Test
+  void check400ResponseStatusWhenTwoKeysCummulateMoreThanMaxRollingPeriodInSameDay() {
+    ResponseEntity<Void> actResponse = executor.executePost(buildPayloadWithKeysThatCummulateMoreThanMaxRollingPeriodPerDay());
```
Comment (pithumke) [inline services/submission/src/test/java/app/coronawarn/server/services/submission/controller/PayloadValidationTest.java:171]: Typo: cumulate

**Review by pithumke — CHANGES_REQUESTED** [review]

File: services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java (hunk 7 above)
Comment (ioangut) [reply services/submission/src/main/java/app/coronawarn/server/services/submission/validation/ValidSubmissionPayload.java:159]: Here the name of the variables are misleading. I will rename them

**Review by pithumke — APPROVED** [review]

**Review by EugenM-SAP — APPROVED** [review]

**Pull request merged**